tests/unit/test_endpoints_v2.py

```diff
@@ -0,0 +1,614 @@
+                        }
+                    ]
+                },
+                headers={
+                    "x-amz-region-set": [
+                        "us-east-1",
+                    ]
+                },
+            ),
+        ),
+    ],
+)
+def test_endpoint_resolution_with_is_valid_host_label(
+    ruleset_dict, input_parameters, expected_endpoint
+):
+    endpoint_provider = EndpointProvider(ruleset_dict)
+    endpoint = endpoint_provider.resolve_endpoint(**input_parameters)
+    assert endpoint.url == expected_endpoint.url
+    assert endpoint.properties == expected_endpoint.properties
+    assert endpoint.headers == expected_endpoint.headers
+
+
+def test_endpoint_resolution_invalid_host_label(ruleset_dict):
+    params = {
+        "Region": "us-east-1",
+        "Bucket": "arn:aws:s3:us-west-2:123456789012:accesspoint:myendpoint",
+        "HostLabel": "foo://this.is.not:.a.\va!lid.host.label",
+    }
+    endpoint_provider = EndpointProvider(ruleset_dict)
+    with pytest.raises(EndpointResolutionError) as exc_info:
+        endpoint_provider.resolve_endpoint(**params)
+    assert str(exc_info.value) == "Default error message"
+
+
+@pytest.mark.parametrize(
+    "label,allow_subdomains,expected_output",
+    [
+        (None, False, False),
+        (None, True, False),
+        ("test.me.please.com", False, False),
+        ("test.me.please.com", True, True),
+        ("s3.us-east-1.amazonaws.com", True, True),
+        ("s3.us-east-1.amazonaws.com", False, False),
+        ("amazon.com", False, False),
+        ("amazon.com", True, True),
+        ("foo", False, True),
+        ("foo", True, True),
+    ],
+)
+def test_is_valid_host_label(label, allow_subdomains, expected_output):
+    assert _is_valid_host_label(label, allow_subdomains) == expected_output
+
+
+def test_partition_none_return_none():
+    assert _aws_partition(None) is None
+
+
+@pytest.mark.parametrize("input_arn", [None, "arn:aws:this-is-not-an-arn:foo"])
+def test_parse_arn_none_return_none(input_arn):
+    assert _aws_parse_arn(input_arn) is None
+
+
+@pytest.mark.parametrize(
+    "uri,expected_output",
+    [
+        (None, None),
+        (
+            ":/?#[]@!$&'()*+,;=%.",
+            "%3A%2F%3F%23%5B%5D%40%21%24%26%27%28%29%2A%2B%2C%3B%3D%25.",
+        ),
+        ("?foo=bar&baz=qux", "%3Ffoo%3Dbar%26baz%3Dqux"),
+    ],
+)
+def test_uri_encode(uri, expected_output):
+    assert _uri_encode(uri) == expected_output
+
+
+def test_parse_url_none_return_none():
+    assert _parse_url(None) is None
+
+
+def test_string_equals_wrong_type_raises():
+    with pytest.raises(RuleSetError) as exc_info:
+        _string_equals(1, 2)
+    assert str(exc_info.value) == (
+        "Input parameters are invalid: Both values must be strings.\n"
+        "value1: 1\n"
+        "value2: 2"
+    )
+
+
+def test_boolean_equals_wrong_type_raises():
+    with pytest.raises(RuleSetError) as exc_info:
+        _boolean_equals(1, 2)
+    breakpoint()
```

Comment:
I think this line slipped through from debugging.

Reply:
it did! i really need to add catching debuggers in my pre-commit config